File: src/main/java/au/id/micolous/metrodroid/transit/ezlink/EZLinkTrip.java
/*
 * EZLinkTrip.java
 *
 * Copyright 2011 Sean Cross <sean@chumby.com>
 * Copyright 2011-2012 Eric Butler <eric@codebutler.com>
 * Copyright 2012 Victor Heng
 * Copyright 2012 Toby Bonang
 * Copyright 2018 Michael Farrell <micolous+git@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.ezlink;

import android.os.Parcel;
import android.support.annotation.Nullable;

import org.jetbrains.annotations.NonNls;

import java.util.Calendar;

import au.id.micolous.farebot.R;
import au.id.micolous.metrodroid.transit.Station;
import au.id.micolous.metrodroid.transit.TransitCurrency;
import au.id.micolous.metrodroid.transit.Trip;
import au.id.micolous.metrodroid.util.Utils;

public class EZLinkTrip extends Trip {
    public static final Creator<EZLinkTrip> CREATOR = new Creator<EZLinkTrip>() {
        public EZLinkTrip createFromParcel(Parcel parcel) {
            return new EZLinkTrip(parcel);
        }

        public EZLinkTrip[] newArray(int size) {
            return new EZLinkTrip[size];
        }
    };
    private final CEPASTransaction mTransaction;
    private final @NonNls String mCardName;

    public EZLinkTrip(CEPASTransaction transaction, String cardName) {
        mTransaction = transaction;
        mCardName = cardName;
    }

    EZLinkTrip(Parcel parcel) {
        mTransaction = parcel.readParcelable(CEPASTransaction.class.getClassLoader());
        mCardName = parcel.readString();
    }

    @Override
    public Calendar getStartTimestamp() {
        return mTransaction.getTimestamp();
    }

    @Override
    public String getAgencyName(boolean isShort) {
        return getAgencyName(mTransaction.getType(), mCardName, isShort);
    }

    @Override
    public String getRouteName() {
        if (mTransaction.getType() == CEPASTransaction.TransactionType.BUS) {
            if (mTransaction.getUserData().startsWith("SVC") || mTransaction.getUserData().startsWith("BUS"))
                return Utils.localizeString(R.string.ez_bus_number,
                        mTransaction.getUserData().substring(3, 7).replace(" ", ""));
            return Utils.localizeString(R.string.unknown_format, mTransaction.getUserData());
        } else if (mTransaction.getType() == CEPASTransaction.TransactionType.BUS_REFUND)
            return Utils.localizeString(R.string.ez_bus_refund);
        else if (mTransaction.getType() == CEPASTransaction.TransactionType.MRT)
            return Utils.localizeString(R.string.ez_mrt);
        else if (mTransaction.getType() == CEPASTransaction.TransactionType.TOP_UP)
            return Utils.localizeString(R.string.ez_topup);
        else if (mTransaction.getType() == CEPASTransaction.TransactionType.CREATION)
            return Utils.localizeString(R.string.ez_first_use);
        else if (mTransaction.getType() == CEPASTransaction.TransactionType.RETAIL)
            return Utils.localizeString(R.string.ez_retail_purchase);
        else if (mTransaction.getType() == CEPASTransaction.TransactionType.SERVICE)
            return Utils.localizeString(R.string.ez_service_charge);
        return Utils.localizeString(R.string.unknown_format, mTransaction.getType().toString());
    }

    @Nullable
    @Override
    public TransitCurrency getFare() {
        if (mTransaction.getType() == CEPASTransaction.TransactionType.CREATION)
            return null;
        return TransitCurrency.SGD(-mTransaction.getAmount());
    }

    @SuppressWarnings("MagicCharacter")
    @Override
    public Station getStartStation() {
        if (mTransaction.getType() == CEPASTransaction.TransactionType.BUS
                && (mTransaction.getUserData().startsWith("SVC")
                || mTransaction.getUserData().startsWith("BUS")))
            return null;
        if (mTransaction.getType() == CEPASTransaction.TransactionType.CREATION)
            return Station.nameOnly(mTransaction.getUserData());
        if (mTransaction.getUserData().charAt(3) == '-'
                || mTransaction.getUserData().charAt(3) == ' ') {
            String startStationAbbr = mTransaction.getUserData().substring(0, 3);
            return EZLinkTransitData.getStation(startStationAbbr);
        }
        return Station.nameOnly(mTransaction.getUserData());
    }

    @SuppressWarnings("MagicCharacter")
    @Override
    public Station getEndStation() {
        if (mTransaction.getType() == CEPASTransaction.TransactionType.CREATION)
            return null;
        if (mTransaction.getUserData().charAt(3) == '-'
                || mTransaction.getUserData().charAt(3) == ' ') {
            String endStationAbbr = mTransaction.getUserData().substring(4, 7);
            return EZLinkTransitData.getStation(endStationAbbr);
        }
        return null;
    }

    public static Mode getMode(CEPASTransaction.TransactionType type) {
        if (type == CEPASTransaction.TransactionType.BUS
                || type == CEPASTransaction.TransactionType.BUS_REFUND)
            return Mode.BUS;
        else if (type == CEPASTransaction.TransactionType.MRT)
            return Mode.METRO;
        else if (type == CEPASTransaction.TransactionType.TOP_UP)
            return Mode.TICKET_MACHINE;
        else if (type == CEPASTransaction.TransactionType.RETAIL
                || type == CEPASTransaction.TransactionType.SERVICE)
            return Mode.POS;
        return Mode.OTHER;
    }


    @Override
    public Mode getMode() {
        return getMode(mTransaction.getType());
    }

    public void writeToParcel(Parcel parcel, int flags) {
        parcel.writeParcelable(mTransaction, flags);
    }

    public int describeContents() {
        return 0;
    }

    public static String getAgencyName(CEPASTransaction.TransactionType type, @NonNls String cardName, boolean isShort) {
        if (type == CEPASTransaction.TransactionType.BUS
                || type == CEPASTransaction.TransactionType.BUS_REFUND) {
            return "BUS";
        }
        if (type == CEPASTransaction.TransactionType.CREATION
                || type == CEPASTransaction.TransactionType.TOP_UP
                || type == CEPASTransaction.TransactionType.SERVICE) {
            if (isShort && cardName.equals("EZ-Link")) return "EZ";
            return cardName;
        }
        if (type == CEPASTransaction.TransactionType.RETAIL) {
            return "POS";
        }
        return "SMRT";
    }

    public static String getRouteName(CEPASTransaction.TransactionType type, String userData) {
        if (type == CEPASTransaction.TransactionType.BUS) {
            if (userData.startsWith("SVC") || userData.startsWith("BUS"))
                return Utils.localizeString(R.string.ez_bus_number,
                        userData.substring(3, 7).replace(" ", ""));
            return Utils.localizeString(R.string.unknown_format, userData);
        } else if (type == CEPASTransaction.TransactionType.BUS_REFUND)
            return Utils.localizeString(R.string.ez_bus_refund);
        else if (type == CEPASTransaction.TransactionType.MRT)
            return Utils.localizeString(R.string.ez_mrt);
        else if (type == CEPASTransaction.TransactionType.TOP_UP)
            return Utils.localizeString(R.string.ez_topup);
        else if (type == CEPASTransaction.TransactionType.CREATION)
            return Utils.localizeString(R.string.ez_first_use);
        else if (type == CEPASTransaction.TransactionType.RETAIL)
            return Utils.localizeString(R.string.ez_retail_purchase);
        else if (type == CEPASTransaction.TransactionType.SERVICE)
            return Utils.localizeString(R.string.ez_service_charge);
        return Utils.localizeString(R.string.unknown_format, type.toString());
    }
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/transit/ezlink/EZLinkTrip.kt
/*
 * EZLinkTrip.kt
 *
 * Copyright 2011 Sean Cross <sean@chumby.com>
 * Copyright 2011-2012 Eric Butler <eric@codebutler.com>
 * Copyright 2012 Victor Heng
 * Copyright 2012 Toby Bonang
 * Copyright 2018 Michael Farrell <micolous+git@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.ezlink

import au.id.micolous.metrodroid.time.Timestamp
import au.id.micolous.metrodroid.multi.FormattedString
import au.id.micolous.metrodroid.multi.R
import au.id.micolous.metrodroid.multi.Localizer
import au.id.micolous.metrodroid.multi.Parcelize
import au.id.micolous.metrodroid.transit.Station
import au.id.micolous.metrodroid.transit.TransitCurrency
import au.id.micolous.metrodroid.transit.Trip

data class EZUserData(val startStation: Station?,
                      val endStation: Station?,
                      val routeName: FormattedString) {
    companion object {
        fun parse(userData: String, type: CEPASTransaction.TransactionType): EZUserData {
            if (type == CEPASTransaction.TransactionType.BUS && (
                            userData.startsWith("SVC")
                                    || userData.startsWith("BUS")))
                return EZUserData(null, null,
                        Localizer.localizeFormatted(R.string.ez_bus_number, userData.substring(3, 7).replace(" ", "")))
            if (type == CEPASTransaction.TransactionType.CREATION)
                return EZUserData(Station.nameOnly(userData), null, Localizer.localizeFormatted(R.string.ez_first_use))
            if (type == CEPASTransaction.TransactionType.RETAIL)
                return EZUserData(Station.nameOnly(userData), null, Localizer.localizeFormatted(R.string.ez_retail_purchase))

            val routeName = when (type) {
                CEPASTransaction.TransactionType.BUS -> Localizer.localizeFormatted(R.string.unknown_format, userData)
                // FIXME: These aren't actually routes...
                CEPASTransaction.TransactionType.BUS_REFUND -> Localizer.localizeFormatted(R.string.ez_bus_refund)
                CEPASTransaction.TransactionType.MRT -> Localizer.localizeFormatted(R.string.ez_mrt)
                CEPASTransaction.TransactionType.TOP_UP -> Localizer.localizeFormatted(R.string.ez_topup)
                CEPASTransaction.TransactionType.SERVICE -> Localizer.localizeFormatted(R.string.ez_service_charge)
                else -> Localizer.localizeFormatted(R.string.unknown_format, type.toString())
            }

            if (userData.length > 6 && (userData[3] == '-' || userData[3] == ' ')) {
                val startStationAbbr = userData.substring(0, 3)
                val endStationAbbr = userData.substring(4, 7)
                return EZUserData(EZLinkTransitData.getStation(startStationAbbr), EZLinkTransitData.getStation(endStationAbbr), routeName)
            }
            return EZUserData(Station.nameOnly(userData), null, routeName)
        }
    }
}

@Parcelize
class EZLinkTrip (private val mTransaction: CEPASTransaction,
                  private val mCardName: String): Trip() {

    override val startTimestamp: Timestamp?
        get() = mTransaction.timestamp

    override val routeName: FormattedString?
        get() = EZUserData.parse(mTransaction.userData, mTransaction.type).routeName

    override val humanReadableRouteID: String?
        get() = mTransaction.userData

    override val fare: TransitCurrency?
        get() = if (mTransaction.type === CEPASTransaction.TransactionType.CREATION) null else TransitCurrency.SGD(-mTransaction.amount)

    override val startStation: Station?
        get() = EZUserData.parse(mTransaction.userData, mTransaction.type).startStation

    override val endStation: Station?
        get() = EZUserData.parse(mTransaction.userData, mTransaction.type).endStation

    override val mode: Trip.Mode
        get() = getMode(mTransaction.type)

    override fun getAgencyName(isShort: Boolean) =
        getAgencyName(mTransaction.type, mCardName, isShort)

    companion object {
        fun getMode(type: CEPASTransaction.TransactionType) =
                when (type) {
                    CEPASTransaction.TransactionType.BUS, CEPASTransaction.TransactionType.BUS_REFUND -> Trip.Mode.BUS
                    CEPASTransaction.TransactionType.MRT -> Trip.Mode.METRO
                    CEPASTransaction.TransactionType.TOP_UP -> Trip.Mode.TICKET_MACHINE
                    CEPASTransaction.TransactionType.RETAIL, CEPASTransaction.TransactionType.SERVICE -> Trip.Mode.POS
                    else -> Trip.Mode.OTHER
                }

        fun getAgencyName(type: CEPASTransaction.TransactionType, cardName: String, isShort: Boolean) = FormattedString.english(
                when (type) {
                    CEPASTransaction.TransactionType.BUS, CEPASTransaction.TransactionType.BUS_REFUND -> "BUS"
                    CEPASTransaction.TransactionType.CREATION,
                    CEPASTransaction.TransactionType.TOP_UP,
                    CEPASTransaction.TransactionType.SERVICE -> if (isShort && cardName == "EZ-Link") "EZ" else cardName
                    CEPASTransaction.TransactionType.RETAIL -> "POS"
                    else -> "SMRT"
                }
            )
    }
}
